New Examples for Posix Threads + Memory Lint Errors

msaxena2 opened the pull request:
WIP. Not yet ready.

msaxena2 commented:
@dwightguth @grosu Please review.

dwightguth commented on examples/simple/Error_Codes.csv:
```diff
@@ -95,6 +95,14 @@ UB-STDLIB1,"Called realloc on memory not allocated by malloc, or already freed."
 UB-STRING1,Null pointer passed to strcpy.,"7.1.4:1, J.2:1 item 108",Undefined Behavior
 UB-STDLIB2,"Called free on memory not allocated by malloc, or already freed.","7.22.3.3:2, J.2:1 item 179",Undefined Behavior
 UB-TTE1,program requires the formation of a composite type from a variable length array type whose size is specified by an expression that is not evaluated.,"6.2.7:3, J.2:1 item 16",Undefined Behavior
+UB-PTHREAD1,Initializing an already initialiized mutex,POSIX 2008 sec. pthread_mutex_init,Undefined Behavior
```
I'd like to see an extra column with like an X or something in it for those errors that are only caught with RV-Match.

dwightguth commented on examples/simple/lint/L-EXC3-bad.c:
```diff
@@ -0,0 +1,20 @@
+int main()
+{
+    void (*func_ptr)(int) = &func;
+    if(func_ptr != ((void *) NULL))
```
I think just NULL should be fine because NULL is usually already of type void *

dwightguth replied:
also I believe the idea behind this error is it occurs with functions that take no arguments when the user has forgotten to add the () that causes the function to actually be invoked. So I think the good example here should be that we are actually calling func().

dwightguth commented on examples/simple/posix/UB-PTHREAD5-bad.c:
```diff
@@ -0,0 +1,17 @@
+int main(void)
+{
+    //Calling pthread_mutex_lock without a call to pthread_mutex_init
+    lock_ptr = malloc(sizeof(pthread_mutex_t));
```
I think to make this more interesting what we should do is to initialize the mutex with pthread_mutex_init, but then copy it with memcpy and call the method on the copied mutex.

dwightguth commented:
looks pretty good. there's a couple things I would do to make this a little more interesting, but otherwise it seems solid to me.

msaxena2 commented:
@dwightguth Please re-review.

dwightguth commented:
looks good. If the tests pass and @grosu is okay with it go ahead and merge this.

grosu commented on examples/simple/lint/L-CEIE1-good.c:
```diff
@@ -0,0 +1,12 @@
+    int * number_ptr = malloc(sizeof(int));
+    *number_ptr = 10;
+    free(number_ptr);
+    printf("No memory Leaks, Execution Successful\n");
```
Be uniform: either capitalize all words or none.  In fact, since the bad example was minimal, you can also make this one minimal (ie, remove assignment and printf).  Or did you have a reason for which you added the additional statements?

msaxena2 replied:
I agree the printf was unnecessary; it's been removed.

grosu commented on examples/simple/lint/L-EXC3-bad.c:
```diff
@@ -0,0 +1,22 @@
+{
+    int * retval = malloc(sizeof(int));
+    *retval = 1;
+    return retval; 
```
Don't we have a L-CEIE1 error/warning here, that memory is malloced but not freed?

dwightguth replied:
No because this function is never actually called.

grosu replied:
I would either remove the malloc statement, or otherwise add a comment in the code saying that the function is not called, so no memory leaks, both in the bad and in the good example. That's because some static analysis tools may report an error here, and we do not want them to think that we "missed" the error.  Something as simple as:

// No memory leak reported, because the function is never invoked.

dwightguth replied:
yeah I'd be inclined to just make the function return a pointer to a statically allocated object, eg:
```
int x;
void *func() { return &x; }
```

dwightguth replied:
and then do that for both bad and good

msaxena2 replied:
Malloc has been replaced by a statically allocated object in both -good and -bad versions.

grosu commented on examples/simple/lint/L-STDLIBE2-good.c:
```diff
@@ -0,0 +1,15 @@
+
+int main()
+{
+    size_t array_size = 20;
```
Does this need to be size_t instead of int?  One would expect the -good programs to only fix the -bad ones minimally.

dwightguth replied:
good catch

grosu replied:
so, do we change this to int or not?

msaxena2 replied:
for now, I've kept size_t in both good and bad versions. We can change it if needed.

grosu commented on examples/simple/lint/L-STDLIBE3-good.c:
```diff
@@ -0,0 +1,15 @@
+
+int main()
+{
+    size_t array_size = 20;
```
same here

grosu commented on examples/simple/posix/UB-PTHREAD2-bad.output:
```diff
@@ -0,0 +1,12 @@
+  at pthread_mutex_lock(UB-PTHREAD2-bad.c:15)
+  at locker_thread(UB-PTHREAD2-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
Is this white line necessary?  I initially thought that there are two error messages.  Others may be confused, too.

dwightguth replied:
@msaxena2 go ahead and remove this line and I will fix the code in the upcoming pull request I am working on that deals with examples.

grosu commented on examples/simple/posix/UB-PTHREAD4-bad.output:
```diff
@@ -0,0 +1,12 @@
+  at pthread_mutex_lock(UB-PTHREAD4-bad.c:15)
+  at locker_thread(UB-PTHREAD4-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
same here: is the whiteline necessary, or follows some more general convention?  if not, i would remove it.

grosu commented on examples/simple/posix/UB-PTHREAD5-good.c:
```diff
@@ -0,0 +1,19 @@
+
+int main(void)
+{
+    
```
any reason for this whiteline?  you typically didn;t have whitelines at the beginning of functions.  also, not clear to me when you put one whiteline and when you put more.  no big deal, but in general it is a good idea to follow consistent formatting and rules, because that shows seriousness and professionalism.

grosu commented on examples/simple/posix/UB-PTHREAD6-bad.output:
```diff
@@ -0,0 +1,7 @@
+Error: UB-PTHREAD6
+Description: Attempting to unlock a PTHREAD_MUTEX_DEFAULT or PTHREAD_MUTEX_NORMAL mutex which is not robust and which has not been locked by the current thread.
```
This description is a bit hard to follow.  Can you remove `which is not robust and`?  Since it is an undefined behavior, it is sort of understood that it is not robust.  Or am I misunderstanding it?

dwightguth replied:
@grosu  a robust mutex is a special type of mutex in POSIX 2008. Attempting to unlock a robust mutex does not result in undefined behavior, hence the error message. If you can think of another way of rephrasing this, I am open to suggestions.

grosu replied:
But the type of the error is Undefined behavior.  If this was a warning and the type potentially undesired behavior, then I would have raised no objections.

dwightguth replied:
It is considered undefined behavior, yes. Specifically it is considered undefined by the POSIX 2008 standard to unlock a mutex you do not own and have locked, unless the mutex is a  robust mutex, in which case it is well defined and returns a specific error code.

grosu replied:
OK then.  What I didn't know was that `robust` is a technical term in this context.  Hopefully others will not get confused.

grosu commented on examples/simple/posix/UB-PTHREAD6-bad.c:
```diff
@@ -0,0 +1,44 @@
+
+void * locker_thread(void * ptr) 
+{
+    //unlocking an unlocked thread
```
I am all for such comments, but we should aim at consistency: either we add such comments to all -bad examples, or to none, and we should either start all of them with a capital letter or none, and we should have a period at the end everywhere or nowhere.  actually I'm not very sure about the first, it kind of makes sense to add such comments selectively, where things are tricky.

grosu commented on examples/simple/posix/UB-PTHREAD8-good.c:
```diff
@@ -0,0 +1,47 @@
+    printf("Successful Execution\n");
+    
+    return 0;
+}
```
The -good one seems a lot larger and different than the -bad one.  Can they get closer to each other?  I think people will like to see quickly the fix.

grosu commented:
I have no other comments.  Very nice examples!  In fact, I think having people go through our canonical examples is a good idea, because they both learn C better and appreciate RV-Match more.  I'm going to add a sentence or two about that in the documentation.  Thank you, guys, very good work!

msaxena2 commented:
@dwightguth @grosu Made some more changes. Please re-review.

grosu commented on examples/simple/lint/L-CCVE2-bad.output:
```diff
@@ -0,0 +1,5 @@
+Warning: L-CCVE2
+Description: conversion to unsigned integer outside the range that can be represented.
```
The description is still not capitalized.

dwightguth commented:
looks good. seems like you addressed all the comments. You tested all the examples and they seem to work as intended? If so, and if @grosu is okay, I am fine with this being merged.

grosu commented on examples/simple/lint/L-CEIE1-bad.c:
```diff
@@ -0,0 +1,10 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
```
You had two lines before the main function until now, and now you have one line.

@msaxena2 , can you please go through my previous comments and address them one by one, telling whether you took care of it and how, or whether you have not and why?  Otherwise I'm wasting a lot of time re-checking everything that I've already reported.

msaxena2 replied:
I've added another whitespace here to adhere to a consistent format, described in one of the later comments.

dwightguth commented on examples/simple/lint/L-CCVE1-bad.output:
```diff
@@ -0,0 +1,5 @@
+Warning: L-CCVE1
+Description: unsigned integer overflow.
```
this and the one @grosu mention should be capitalized. I will make a PR to handle the code change.

grosu commented on examples/simple/lint/L-EXC3-bad.c:
```diff
@@ -0,0 +1,21 @@
+    if(func_ptr != NULL)
+    {
+       printf("Hello, World!\n"); 
+        
```
why a white line here?

grosu commented on examples/simple/lint/L-PTHREAD9-bad.output:
```diff
@@ -0,0 +1,8 @@
+  at pthread_mutex_lock(L-PTHREAD9-bad.c:15)
+  at locker_thread(L-PTHREAD9-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
why whiteline?

grosu commented on examples/simple/lint/L-PTHREAD9-good.c:
```diff
@@ -0,0 +1,44 @@
+#include<pthread.h>
+#include<stdlib.h>
+
+
```
why 3 whitelines?

msaxena2 replied:
Changed to 2 whitespaces.

grosu commented on examples/simple/lint/L-STDLIBE2-good.c:
```diff
@@ -0,0 +1,15 @@
+    int * small_int_array = malloc(array_size * sizeof(int));
+    if(small_int_array != NULL)
+    {
+    
```
why whiteline?

msaxena2 replied:
Removed the whitespace here.

grosu commented on examples/simple/lint/L-STDLIBE3-bad.c:
```diff
@@ -0,0 +1,15 @@
+
+int main()
+{
+    size_t array_size = 1024 * 1024 * 1024;
```
int?

grosu commented on examples/simple/posix/UB-PTHREAD1-bad.c:
```diff
@@ -0,0 +1,24 @@
+
+int main(void)
+{
+
```
why whiteline?

msaxena2 replied:
Addressed by removing the whiteline.

grosu commented on examples/simple/posix/UB-PTHREAD6-bad.c:
```diff
@@ -0,0 +1,43 @@
+        return 1;
+    }
+
+
```
3 whitelines?

msaxena2 replied:
Addressed by removing the whiteline.

msaxena2 commented:
@grosu @dwightguth Please re-review. Changes made since previous review:
* I've addressed all the whiteline related comments that had been made in the previous review. The files now have 2 whitelines between the last '#include' and the following line of code, and a single whiteline between different function definitions. Within function blocks, I've removed random occurrences of whitelines. I've done this for all files to ensure uniformity.
* Both good and bad version of STDLIBE2, STDLIBE3 now use size_t instead of int. I'm open to reverting them back to int.

grosu commented:
lgtm
